// @node-red/nodes/core/neural/10-neural-input.js
module.exports = function(RED) {
    "use strict";

    var neuralStepMode = false;
    var neuralSpeed = 4;

    RED.neural = RED.neural || {};
    RED.neural.outputs = {};
    RED.neural.connections = {};

    function trackWires(node) {
        var wires = node.wires || [];
        RED.neural.connections[node.id] = [];
        wires.forEach(function(wireGroup) {
            wireGroup.forEach(function(targetId) {
                var target = RED.nodes.getNode(targetId);
                if (target) {
                    RED.neural.connections[node.id].push({
                        id: targetId,
                        type: target.type,
                        name: target.name
                    });
                }
            });
        });
    }

    function NeuralInputNode(n) {
        RED.nodes.createNode(this, n);
        var node = this;
        node.values = n.values || "0.5, 0.5, 0.5";
        if (typeof node.values === "string") {
            node.values = node.values.split(",").map(function(v) { return parseFloat(v.trim()) || 0; });
        }
        node.count = Number(n.count) || 3;

        trackWires(node);

        node.on("input", function(msg, send, done) {
            var vals = node.values.slice();
            while (vals.length < node.count) {
                vals.push(0);
            }
            vals.length = node.count;
            msg.payload = vals;
            msg.neural = msg.neural || {};
            msg.neural.type = "input";
            msg.neural.shape = [vals.length];
            msg.neural.step = (msg.neural.step || 0) + 1;

            var active = vals.filter(function(v) { return v > 0; }).length;
            node.status({ fill: active > 0 ? "green" : "grey", shape: "dot", text: vals.map(function(v) { return v.toFixed(2); }).join(", ") });

            RED.neural.outputs[node.id] = { values: vals, type: "neural-input", name: node.name || "Input", timestamp: Date.now(), shape: [vals.length] };

            send(msg);
            done();
        });
    }

    RED.nodes.registerType("neural-input", NeuralInputNode);

    RED.httpAdmin.post("/neural-inject/:id", RED.auth.needsPermission("inject.write"), function(req, res) {
        var node = RED.nodes.getNode(req.params.id);
        if (node != null) {
            try {
                var msg = req.body || {};
                node.receive(msg);
                res.sendStatus(200);
            } catch (err) {
                res.sendStatus(500);
                node.error(err.toString());
            }
        } else {
            res.sendStatus(404);
        }
    });

    RED.httpAdmin.post("/neural-advance", RED.auth.needsPermission("inject.write"), function(req, res) {
        RED.nodes.eachNode(function(node) {
            if (node.neuralRelease) {
                node.neuralRelease();
            }
        });
        res.sendStatus(200);
    });

    RED.httpAdmin.post("/neural-speed", RED.auth.needsPermission("inject.write"), function(req, res) {
        if (req.body && req.body.speed !== undefined) {
            neuralSpeed = req.body.speed;
        }
        res.sendStatus(200);
    });

    RED.httpAdmin.get("/neural-speed", function(req, res) {
        res.json({ speed: neuralSpeed });
    });

    RED.httpAdmin.get("/neural-state", function(req, res) {
        var outputs = {};
        Object.keys(RED.neural.outputs).forEach(function(id) {
            var node = RED.nodes.getNode(id);
            if (node) {
                outputs[id] = RED.neural.outputs[id];
                outputs[id].active = true;
            }
        });

        RED.nodes.eachNode(function(node) {
            if (node.type && node.type.indexOf("neural-") === 0) {
                if (!outputs[node.id]) {
                    trackWires(node);
                    outputs[node.id] = {
                        values: [],
                        type: node.type,
                        name: node.name || "",
                        shape: [],
                        timestamp: 0,
                        active: false,
                        status: node._held ? "held" : "idle"
                    };
                }
                if (!outputs[node.id].connections) {
                    outputs[node.id].connections = RED.neural.connections[node.id] || [];
                }
                if (node._held) {
                    outputs[node.id].status = "held";
                } else if (!outputs[node.id].active) {
                    outputs[node.id].status = "idle";
                } else {
                    outputs[node.id].status = "active";
                }
            }
        });

        res.json(outputs);
    });

    RED.httpAdmin.post("/neural-reset", function(req, res) {
        RED.neural.outputs = {};
        res.sendStatus(200);
    });
};
