// @node-red/nodes/core/neural/60-neural-output.js
module.exports = function(RED) {
    "use strict";

    function NeuralOutputNode(n) {
        RED.nodes.createNode(this, n);
        var node = this;
        node.lastOutput = [];

        node.on("input", function(msg, send, done) {
            var input = msg.payload;
            if (!Array.isArray(input)) {
                input = [input];
            }

            node.lastOutput = input;
            node.status({ fill: "green", shape: "dot", text: input.map(function(v) { return v.toFixed(3); }).join(", ") });

            var data = {
                type: "neural-output",
                name: node.name || "Output",
                values: input,
                shape: msg.neural ? msg.neural.shape : [input.length],
                source: msg.neural ? msg.neural.type : "unknown"
            };

            node.send({ payload: input, neural: Object.assign({}, msg.neural), output: data });

            done();
        });
    }

    RED.nodes.registerType("neural-output", NeuralOutputNode);
};
